lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+//===----------------------------------------------------------------------===//
+// Reset Network
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ResetNode;
+
+enum class ResetKind { Uninferred, Async, Sync };
+
+/// A driver association between two IR values, given as a "source" value being
+/// driven onto a "destination" port or wire. Also contains location information
+/// for error reporting to the user.
+struct ResetDrive {
+  /// The output being driven.
+  ResetNode *dst;
+  /// The input node to the drive.
+  ResetNode *src;
+  /// The location to use for diagnostics.
+  Location loc;
+};
+
+/// A collection of IR values (represented by their `ResetNode`) that are
+/// connected together through instance ports or explicit connect ops. This
+/// provides a data structure that records all the ports and wires in a design
+/// that are transitively connected, and which thus must carry the same reset
+/// type.
+///
+/// While building the `ResetMap`, there will be a lot of small transient
+/// `ResetNet`s for the initial connections to a wire or port. As more
+/// connections are added, more and more of these `ResetNet`s are combined into
+/// larger nets. Since a design only has a handful of resets, the number of
+/// final `ResetNet`s is expected to be low.
+struct ResetNet {
+  /// The nodes in this reset network. Each node corresponds to a value in the
+  /// IR that is either used as the LHS or RHS of a connection to this network.
+  SmallSetVector<ResetNode *, 16> nodes;
+
+  /// The drives that contribute to this network.
+  SmallVector<ResetDrive, 32> drives;
+
+  /// The inferred kind of the reset.
+  ResetKind kind = ResetKind::Uninferred;
+
+  void clear();
+  FieldRef guessRoot();
+};
+
+/// Metadata associated with a single IR value (represented by a `FieldRef`),
+/// and the reset network the value belongs to. Created on-demand by `ResetMap`
+/// whenever `getNode` encounters a new value.
+struct ResetNode {
+  /// The value in the IR that corresponds to this node.
+  FieldRef value;
+  /// The type of the value.
+  FIRRTLType type;
+  /// The reset net this node belongs to.
+  ResetNet *net = nullptr;
+};
+
+/// A global view of all reset networks in a design.
+///
+/// A `ResetMap` associates values in a design with a `ResetNet`. It does this
+/// by creating a `ResetNode` for every value (represented by a `FieldRef`),
+/// which holds a pointer to the reset network the value belongs to. The core
+/// functionality is the `add` function, which stores a connection between two
+/// values in the map. Doing so associates a "source" as the value being driven
+/// onto a "destination" port or wire, and combines the `ResetNet` the values
+/// belong to into a single one. This transitively establishes a `ResetNet` as
+/// the collection of all values in the IR that must have the same reset type,
+/// since they are connected together (through instance ports or connect ops).
```

Comment:
I don't really understand the core algorithm here, but would a "Union Find" data structure make sense here?  It is an efficient algorithm for building equivalence sets in inverse ackerman time.  There is an implementation in `llvm/ADT/EquivalenceClasses.h`.

Reply:
That looks exactly like what's needed here! Will give it a shot.